sources/engine/Stride.Graphics/Vulkan/DescriptorSetLayout.Vulkan.cs

```diff
@@ -89,21 +89,21 @@ internal static unsafe VkDescriptorSetLayout CreateNativeDescriptorSetLayout(Gra
                         // Remember this, so we can choose the right VkDescriptorType in DescriptorSet.SetShaderResourceView
                         immutableSamplers[i] = entry.ImmutableSampler.NativeSampler;
                         //bindings[i].VkDescriptorType = VkDescriptorType.CombinedImageSampler;
-                        bindings[usedBindingCount].pImmutableSamplers = immutableSamplersPointer + i;
+                        bindings[usedBindingCount].PImmutableSamplers = immutableSamplersPointer + i;
                     }
 
-                    typeCounts[(int)bindings[usedBindingCount].descriptorType] += bindings[usedBindingCount].descriptorCount;
+                    typeCounts[(int)bindings[usedBindingCount].DescriptorType] += bindings[usedBindingCount].DescriptorCount;
 
                     usedBindingCount++;
                 }
 
-                var createInfo = new VkDescriptorSetLayoutCreateInfo
+                var createInfo = new DescriptorSetLayoutCreateInfo
                 {
-                    sType = VkStructureType.DescriptorSetLayoutCreateInfo,
-                    bindingCount = (uint)usedBindingCount,
-                    pBindings = usedBindingCount > 0 ? (VkDescriptorSetLayoutBinding*)Core.Interop.Fixed(bindings) : null,
+                    SType = StructureType.DescriptorSetLayoutCreateInfo,
+                    BindingCount = (uint)usedBindingCount,
+                    PBindings = usedBindingCount > 0 ? (DescriptorSetLayoutBinding*)Core.Interop.Fixed(bindings) : null,
                 };
-                vkCreateDescriptorSetLayout(device.NativeDevice, &createInfo, null, out var descriptorSetLayout);
+                GetApi().CreateDescriptorSetLayout(device.NativeDevice, &createInfo, null, out var descriptorSetLayout);
```

Comment:
Note: GetApi is an extremely expensive call, please don't use it every time you use an API function!

The `Vk` object stores all function pointers, which are lazily initialized (i.e. the first time you use a function on that object, there's a small penalty in which we call `vkGetInstanceProcAddr`, `vkGetDeviceProcAddr`, or the kernel `GetProcAddress` method). If you make a new object for every call, we call this every time and incur this penalty every time. Moreover, there's a few allocations that happen under-the-hood that are intended to be long-lived when instantiating the object, and once again are pretty expensive.

As a general rule of thumb, you should have one Vk instance for each instance-device combo. This ensures you load the correct function pointers for each device, and don't encounter any indirection in the Vulkan driver in multi-GPU scenarios.

Usually this means one object for a whole program, but with Stride it doesn't look to be as simple. It looks like:
- create an object in [GraphicsAdapterFactory.InitializeInternal](https://github.com/stride3d/stride/blob/7e836297cb5930c01e6dfa0183e7f3cc64748fb6/sources/engine/Stride.Graphics/Vulkan/GraphicsAdapterFactory.Vulkan.cs#L24)
- add an argument to the `GraphicsAdapterFactoryInstance` constructor for the Vk object. make it nullable and default to null, and add logic such that if not provided it just creates a new API object (this is primarily for the `debugInstance`)
- when constructing the `defaultInstance`, pass this Vk object into the constructor. `defaultInstance` can now be thought of as "owning" that API object, and when `vkCreateInstance` is called it will capture your instance handle implicitly and rebuild the function table such that it calls `vkGetInstanceProcAddr` with that handle.
- `GraphicsDevice` should also hold a Vk object. In [GraphicsDevice.InitializePlatformDevice](https://github.com/stride3d/stride/blob/5bb8915ccc596c5bd0a404b5f59ed61031a9251e/sources/engine/Stride.Graphics/Vulkan/GraphicsDevice.Vulkan.cs#L248), do something like:
  ```cs
  // TODO clean this up when https://github.com/dotnet/Silk.NET/issues/697 is a thing
  if (_vk is null)
  {
      _vk = Vk.GetApi();
      _vk.CurrentInstance = GraphicsAdapterFactory.GetInstance(false).NativeInstance;
  }
  ```
- perhaps find some way to have a `Vk` property in [GraphicsResourceBase](https://github.com/stride3d/stride/blob/5bb8915ccc596c5bd0a404b5f59ed61031a9251e/sources/engine/Stride.Graphics/Vulkan/GraphicsResourceBase.Vulkan.cs) to make code look nicer.

This is just my finger-in-the-air estimate of how best to work this into the engine, and is by no means "the best way" (as I'm not that familiar with Stride's internals), but basically it boils down to:
- have one Vk object for each instance-device combo.
- pass it around to the objects that need it.

There are hackier ways to work this in, for example have a static class containing the Vk object, but the above is the best way to account for multiple logical devices.